### ISSUE TITLE
Long training time [Model Search]

### ISSUE THREAD

**Issue by maaft** [ashaw596/squeezenas#6]
Hi,

I've implemented the model search on my own since it is not available yet in your github repo.

Here is my SuperNet-Code of the encoder for the XLarge Model. Note, that I ommited the classifier and `last_conv` layer for simplicity.
```python
class SqueezeNASSuperNet(nn.Module):
    def __init__(self, input_channels, hyperparams: InverseResidualMetaNetHyperparameters):
        super().__init__()
        self.hyperparams = hyperparams
        self.conv1 = conv3x3_bn(c_in=input_channels, c_out=hyperparams.init_channels, stride=2)
        self.residuals: nn.ModuleList[nn.ModuleList] = nn.ModuleList()

        gene_i = 0
        last_c = hyperparams.init_channels

        num_layers = 0
        for block_hyper in hyperparams.blocks:
            for i in range(block_hyper.num_repeat):
                if i == 0:
                    stride = block_hyper.stride
                else:
                    stride = 1
                ops = nn.ModuleList([InvertedResidual(op, last_c, block_hyper.num_channels, stride) for op in Ops])
                self.residuals.append(ops)
                gene_i += 1
                last_c = block_hyper.num_channels
                num_layers += 1

        self.alphas = nn.Parameter(1. / len(Ops) * torch.ones(num_layers, len(Ops)), requires_grad=True)

        self.register_buffer("active_cells", torch.ones(num_layers, len(Ops)))

    def forward(self, inputs, gt=None):
        cur_feat = inputs
        cur_feat = self.conv1(cur_feat)
        residuals_outputs = []

        softmax_alphas = F.softmax(self.alphas, dim=-1)
        if self.training:
            alphas = gumbel_softmax(F.log_softmax(self.alphas), hard=True)
        else:
            idx = F.softmax(self.alphas, dim=-1).argmax(dim=-1, keepdim=True)
            alphas = torch.zeros_like(self.alphas)
            alphas.scatter_(1, idx, 1)

        latency = 0
        for i, (supercell, cell_alphas, active_cells) in enumerate(zip(self.residuals, alphas, self.active_cells)):
            next_feat = 0

            for j, (op, alpha, active) in enumerate(zip(supercell, cell_alphas, active_cells)):
                if not self.training and alpha == 0:
                    continue
                if self.training and softmax_alphas[i, j] <= 0.005 and self.active_cells[i].sum() > 1:
                    self.active_cells[i, j] = 0
                    continue
                if self.training and active == 0:
                    continue
                next_feat += alpha * op(cur_feat)
                latency += alpha * op.forward_latency(cur_feat)
            cur_feat = next_feat
            residuals_outputs.append(cur_feat)

        return {'output': cur_feat, 'residuals_outputs': residuals_outputs, "latency": latency}
```
In a small test I tested my code with using only two Ops, e.g.:
```python
class Ops(enum.Enum):
    inverse_residual_k3_e1_g1 = InvertedResidualOperation(group=1, kernel_size=3, padding=1, expand_ratio=1)

    residual_skipish = InvertedResidualSkipOperation()

    def __repr__(self):
        return super().__str__()
```

I trained on a custom dataset using a batch-size of 8. First I followed your suggestion to use SGD with lr=0.05 but it didn't seem to converge and I switched to Adam with lr=3e-4.
Now this (very simple) test took almost 6 hours to arrive at more or less decent `alphas` and I was wondering what could be the reason that the model search is so slow?

Also I disabled the latency loss as I observed that this will skew the `alphas` to favor the skip-op and hinder learning of useful features of the InvertedResidualOp.

Thanks!

**Comment by ashaw596** [comment]
Hi! What was the learning rate were you using for the architecture and weight parameters? Empirically I found that it necessary to use a seperate learning rate for the architecture parameters and the weight parameters. Also while unforunately, while I hope to release all the code soon, I am unsure when I will be able to due to work circumstances, I have released some a basic supernet utility here https://github.com/ashaw596/supernet-nas Also I might suspect that you weighted the latency loss too heavily if you were seeing a very heavy skew and some bias would be expected if your network is choosing between two networks. Depending on number of gpus, 6 hours isn't terribly unreasonable depending on network size. But you can always force faster architecture parameter convergence rate by increasing the architecture learning rate.

I might also recommend annealing the gumbel softmax temperature as that helps with the stability of the gradients at the start.

**Comment by maaft** [comment]
Thank you for your fast answer!
Could you share your learning-rates for architecture and weight parameters? As you suggested I did use the same learning rate for both.

For now I just weighted my latency-loss with 0.01 (e.g. loss = seg_loss + 0.01 * lat_loss). The skew was not too heavy but noticeable.

I trained the model on a single RTX 2080 Ti but can use up to 8 if I wanted to.

I am wondering though if a higher batch-size actually helps for faster convergence, as the gumbel-softmax-sampling is not done per batch-idx anyway. Or should I rewrite my code such that each image of the batch takes a different path through the super-network?

Or is Multi-GPU training actually **required** for this method as each process on every GPU would sample its own architecture parameters and therefore at each iteration #num-gpus possible network paths are evaluated?

By the way, in another Paper (FasterSeg) they didn't use any sampling for searching the super-cell and used plain softmax instead. I think that this could be problematic since the optimizer learns an optimal weighted **combination** of feature-maps for each super-cell and not an optimal feature-map per operation. Do you agree?

**Comment by maaft** [comment]
Another thing i noticed is that the architecture parameters in the "middle" of the network seem to converge very slow in comparison to parameters at the start or the end of the network.

Here you see active operations. Similar to you, I disable operations when their sample-probability drops below a certain threshold.
```
       op1, op2, op3, ......                               op13
       [[0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],     # layer1
        [0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],     # layer2
        [0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0.],
        [1., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 0.],
        [1., 0., 1., 1., 0., 0., 1., 1., 0., 0., 0., 0., 0.],
        [1., 0., 1., 1., 1., 0., 1., 1., 0., 0., 0., 0., 0.],
        [1., 0., 1., 1., 0., 0., 1., 1., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 1., 0., 1., 1., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 1., 1., 0., 0., 0., 1., 0.]]     # layer22
```

Do you have any idea why this might happen?

**Comment by ashaw596** [comment]
Yes I would definitely recommend doing image level sampling like in my utility. It will greatly decrease the variance an bias in the gradient updates due to the randomness.

**Comment by maaft** [comment]
Thank you, you helped me a lot!
